src/Symfony/Component/HttpClient/Response/AmpResponse.php

```diff
@@ -88,29 +94,33 @@ public function __construct(AmpClientState $multi, Request $request, array $opti
             $onProgress((int) $info['size_download'], ((int) (1 + $info['download_content_length']) ?: 1) - 1, (array) $info);
         };
 
-        $this->id = $id = Loop::defer(static function () use ($request, $multi, &$id, &$info, &$headers, $canceller, &$options, $onProgress, &$handle, $logger) {
-            return self::generateResponse($request, $multi, $id, $info, $headers, $canceller, $options, $onProgress, $handle, $logger);
+        $pauseDeferred = new Deferred();
+        $pause = new Success();
+
+        $throttleWatcher = null;
+
+        $this->id = $id = self::$nextId++;
+        Loop::defer(static function () use ($request, $multi, &$id, &$info, &$headers, $canceller, &$options, $onProgress, &$handle, $logger, &$pause) {
+            return new Coroutine(self::generateResponse($request, $multi, $id, $info, $headers, $canceller, $options, $onProgress, $handle, $logger, $pause));
         });
 
-        $info['pause_handler'] = static function (float $duration) use ($id, &$delay) {
-            if (null !== $delay) {
-                Loop::cancel($delay);
-                $delay = null;
+        $info['pause_handler'] = static function (float $duration) use (&$throttleWatcher, &$pauseDeferred, &$pause) {
+            if (null !== $throttleWatcher) {
+                Loop::cancel($throttleWatcher);
             }
 
-            if (0 < $duration) {
-                $duration += microtime(true);
-                Loop::disable($id);
-                $delay = Loop::defer(static function () use ($duration, $id, &$delay) {
-                    if (0 < $duration -= microtime(true)) {
-                        $delay = Loop::delay(ceil(1000 * $duration), static function () use ($id) { Loop::enable($id); });
-                    } else {
-                        $delay = null;
-                        Loop::enable($id);
-                    }
-                });
+            $pause = $pauseDeferred->promise();
```

Comment:
@nicolas-grekas  I guess this line should be below the following `if`, no?

Reply:
You wrote it this way, and I'm not sure it should be moved below. Can you please confirm?